### PR TITLE
ic-proxy: enable on master pipelines

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
- compile gpdb with ic-proxy enabled on master pipelines;
- add the ICW jobs of ic-proxy mode;
- enable ic-proxy on Travis;
- reduce ic-proxy log level;

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: concourse/pipelines/templates/gpdb-tpl.yml [hunk 1]
```diff
@@ -478,6 +486,13 @@ resources:
     json_key: ((concourse-gcs-resources-service-account-key))
     regexp: gp-internal-artifacts/centos7/libquicklz-devel-(1\.5\.0-.*)-1.el7.x86_64.rpm
 
+- name: libuv-centos7
+  type: gcs
+  source:
+    bucket: ((gcs-bucket))
+    json_key: ((concourse-gcs-resources-service-account-key))
+    regexp: gp-internal-artifacts/centos7/libuv-(1\.18\.0.*).tar.gz
```
Comment (zhangh43) [inline concourse/pipelines/templates/gpdb-tpl.yml:494]: Do we need libuv artifact? I think it should be installed in the docker image.
Also the libuv version supplied by releng team is 1.38.0, but I think we could just use the one in the docker image.

File: concourse/pipelines/templates/gpdb-tpl.yml [hunk 2]
```diff
@@ -937,6 +964,44 @@ jobs:
...
+      - get: bin_gpdb
+        resource: bin_gpdb_centos7
+        passed: [compile_gpdb_centos7]
+        trigger: [[ test_trigger ]]
+      - get: gpdb7-centos7-test
+  - task: ic_gpdb
+    file: gpdb_src/concourse/tasks/ic_gpdb.yml
+    image: gpdb7-centos7-test
+    params:
+      MAKE_TEST_COMMAND: -k PGOPTIONS='-c gp_interconnect_type=proxy -c optimizer=on' EXTRA_REGRESS_OPTS=--exclude-tests=external_table,copydml installcheck-world
```
Comment (zhangh43) [inline concourse/pipelines/templates/gpdb-tpl.yml:982]: you skip external_table and copydml test in ic-proxy, are they test case issues and feature not supported issues? If it's feature not supported issue, we need record them in the document.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 1 above)
Comment (zhangh43) [reply concourse/pipelines/templates/gpdb-tpl.yml:494]: Could you confirm the libuv artifact? Comments from Kris:
```
The builds are available in GCS with a filename Convention of “libuv-1.38.0.build.[N].tar”, such that [N] is a serial build number:
- gs://pivotal-gpdb-concourse-resources-prod/gp-internal-artifacts/centos7/
- gs://pivotal-gpdb-concourse-resources-prod/gp-internal-artifacts/centos6/
- gs://pivotal-gpdb-concourse-resources-prod/gp-internal-artifacts/ubuntu18.04/
- GCP Project: data-gpdb-ci
```

**Comment by pivotal-ning-yu** [comment]
Pushed a commit to change the GUC `gp_interconnect_proxy_addresses` from a developer option to a normal one, and now we are able to reload this GUC with SIGHUP by executing `gpstop -u`.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 1 above)
Comment (pivotal-ning-yu) [reply concourse/pipelines/templates/gpdb-tpl.yml:494]: Aha, libuv 1.38 was uploaded to GCP yesterday.  I will update the PR to use this new one.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 2 above)
Comment (pivotal-ning-yu) [reply concourse/pipelines/templates/gpdb-tpl.yml:982]: They are tracked in an internal story.

**Comment by pivotal-ning-yu** [comment]
Upgraded libuv from 1.18.0 to 1.38.0 .

File: src/backend/cdb/motion/README.ic-proxy.md [hunk 3]
```diff
@@ -606,18 +606,18 @@ the proxy-proxy connections unconditionally.
...
+1. the cluster expansion (adding new segments to the cluster);
+2. the data expansion (expand the data to the new segments);
 
-1. let the `gpexpand` utility updates the `gp_interconnect_proxy_addresses`
-   automatically;
-2. let the `gp_interconnect_proxy_addresses` be able to be reloaded without a
-   cluster restart;
+When ic-proxy is enabled we need a stage 1.5 between them:
+
+1.1. update the GUC `gp_interconnect_proxy_addresses` to include the new
```
Comment (zhangh43) [inline src/backend/cdb/motion/README.ic-proxy.md:617]: %s/1.1/1.5

File: src/backend/utils/misc/guc_gp.c [hunk 4]
```diff
@@ -4360,7 +4360,7 @@ struct config_string ConfigureNamesString_gp[] =
 
 #ifdef ENABLE_IC_PROXY
 	{
-		{"gp_interconnect_proxy_addresses", PGC_POSTMASTER, DEVELOPER_OPTIONS,
+		{"gp_interconnect_proxy_addresses", PGC_SIGHUP, GP_ARRAY_TUNING,
```
Comment (zhangh43) [inline src/backend/utils/misc/guc_gp.c:4363]: GP_ARRAY_TUNING  or GP_ARRAY_CONFIGURATION? This GUC more like a configuration.

File: src/backend/cdb/motion/README.ic-proxy.md (hunk 3 above)
Comment (pivotal-ning-yu) [reply src/backend/cdb/motion/README.ic-proxy.md:617]: LOL, good point.

File: src/backend/utils/misc/guc_gp.c (hunk 4 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/misc/guc_gp.c:4363]: `gp_interconnect_type` is also `GP_ARRAY_TUNING`, but I'm also fine with `GP_ARRAY_CONFIGURATION`. I could change this in next push.

**Comment by pivotal-ning-yu** [comment]
Thanks for the approval.  I rebased to latest master and squashed the fixup commits.  Now I'll monitor the devel pipeline, once it's green I'll merge the PR.

File: src/backend/utils/misc/guc_gp.c (hunk 4 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/misc/guc_gp.c]: Changed to `GP_ARRAY_CONFIGURATION`.

File: src/backend/cdb/motion/README.ic-proxy.md (hunk 3 above)
Comment (pivotal-ning-yu) [reply src/backend/cdb/motion/README.ic-proxy.md]: Changed.

**Comment by pivotal-ning-yu** [comment]
Hmm, a bad news is libuv 1.38 is only in the gcp prod bucket, the dev bucket only has 1.18, so the PR does not compile on devel pipelines at all.  I'll try to copy the libuv 1.38 packages from prod bucket to dev bucket.

**Comment by pivotal-ning-yu** [comment]
I do not have permission to copy packages from prod bucket to dev bucket.  In such a case I have to adjust my plan:
1. wait for Travis builds to pass;
2. merge the PR;
3. monitor the master pipelines;

If there is anything wrong I'll revert the patches.

**Comment by pivotal-ning-yu** [comment]
In fact another option is to convert this PR back to use libuv 1.18, which is available on both prod & dev buckets, and it was verified on devel pipeline.  But I think we could first try with the plan with 1.38.

**Comment by pivotal-ning-yu** [comment]
OK, the Travis CI is green, I'll merge it now.

**Pull request merged**